### PR TITLE
fix(connlib): track post-NAT source port in flow logs

### PR DESCRIPTION
**Pull request by jamilbk**
On the Gateway, `record_translated_packet` picked up the translated destination IP but not the translated source port. When several proxy IPs map to one real IP (name-based virtual hosting), the NAT rewrites the source port to keep outbound flows distinct, so the outbound flow key and the return path disagreed:

| field | outbound key | return path |
|---|---|---|
| dst_ip | real ✓ | real ✓ |
| dst_port | 443 ✓ | 443 ✓ |
| src_port | pre-NAT (54321) | post-NAT (54322) |

Because the keys no longer matched, affected flows recorded no inbound traffic (`rx` stayed at zero), missed the TCP FIN/FIN close, and terminated only on the idle timeout.

The flow key now uses the post-NAT source port, matching what the return path observes.

### CODE REVIEW

**Comment by jamilbk** [comment]
Found this while I was working on reporting flow logs (and preventing collisions) to the portal

**Review by thomaseizinger — APPROVED** [review]
Would be great to extend the integration tests to cover this.

**Comment by thomaseizinger** [comment]
> Would be great to extend the integration tests to cover this.

I meant the docker-based integration tests. That is where we capture flow log functionality right now.

File: rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs [hunk 1]
```diff
@@ -1050,6 +1051,88 @@ mod tests {
     use std::net::Ipv4Addr;
 
     use super::*;
+    use ip_packet::make::{self, TcpFlags};
```
Comment (jamilbk) [inline rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: @thomaseizinger Updating the sim gateway to report flow state so we can model it for integration tests seems to be non-trivial. This regression test is also fairly complex but simpler than the above.

Let me know which direction you prefer to handle testing here.

Since the FlowTracker is already an isolated module having some unit tests for specific boundary conditions might not be a bad idea if there are helpers to make these less verbose?

Comment (thomaseizinger) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: The reason we don't have module integration tests here is because the actual logic of this module is pretty trivial and it working properly depends largely on it being used correctly. Asserting its use correctly in turn depends on piping entire network connections through connlib which we only do in our docker-based integration tests. I don't think test infrastructure for simulating entire network connections in-process carries its weight here. That is why we assert the flow logs in the docker tests by looking at the actual logs that we emit.

A virtual host test case should trigger the port-overlap, right? Such a test case can then also be used to assert the flow logs.

Comment (jamilbk) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: > because the actual logic of this module is pretty trivial

Not trivial enough to be bug-free unfortunately ;-)

> Asserting its use correctly in turn depends on piping entire network connections through connlib

We are just trying to assert the module behaves correctly, no? Not that it's being used correctly? I don't think this change affects how any consumers are using the module right?

Comment (jamilbk) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: Integration tests are not cheap in our system and are already flaky enough so I want to make sure that they are used to test the actual integration between components and not as a crutch for missing unit tests.

In this case if we already have a flow integration test I could update that one.

Comment (thomaseizinger) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: > > because the actual logic of this module is pretty trivial
> 
> Not trivial enough to be bug-free unfortunately ;-)

Yep, every testing strategy is fundamentally a trade-off between complexity and risk-reduction.

> > Asserting its use correctly in turn depends on piping entire network connections through connlib
> 
> We are just trying to assert the module behaves correctly, no? Not that it's being used correctly? I don't think this change affects how any consumers are using the module right?

This is one of these modules where, due to its nature of needing to hook into the packet processing path, it has a very "chatty" API, i.e. it cannot encapsulate a larger piece of functionality behind an abstraction that has meaningful depth. Hence, chances are that any test you write just asserts the current behaviour but does not catch bugs because you are operating on the same abstraction level as the implementation. In my opinion, such tests do not carry their weight. It is like testing that a setter on an object did indeed set the property. Sure, you are testing something. Does it carry its weight? Unlikely.

> Integration tests are not cheap in our system and are already flaky enough so I want to make sure that they are used to test the actual integration between components and not as a crutch for missing unit tests.

Agreed. I do think in this case, extending the integration tests is the correct place to test this behaviour.

The next best alternative IMO would be to change our proptest-infrastructure to simulate a real network stack and run entire TCP connections through it and assert the captured flow logs as a result. We don't have that capability today and it would be a larger undertaking which is why it hasn't happened yet.

Comment (jamilbk) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: 👍

Comment (jamilbk) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: Will update existing flowtracker integration test to cover this

Comment (thomaseizinger) [reply rust/libs/connlib/tunnel/src/gateway/flow_tracker.rs]: Thank you!
